modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCachePreloader.java

```diff
@@ -74,10 +75,13 @@
     /**
      * @param exchId Exchange ID.
      * @param exchFut Completed exchange future. Can be {@code null} if forced or reassigned generation occurs.
+     * @param historicalExclusions Set of node ids that should not be used for historical rebalance.
      * @return Partition assignments which will be requested from supplier nodes.
      */
-    @Nullable public GridDhtPreloaderAssignments generateAssignments(GridDhtPartitionExchangeId exchId,
-                                                                     @Nullable GridDhtPartitionsExchangeFuture exchFut);
+    @Nullable public GridDhtPreloaderAssignments generateAssignments(
+        GridDhtPartitionExchangeId exchId,
+        @Nullable GridDhtPartitionsExchangeFuture exchFut,
+        Set<UUID> historicalExclusions);
```

Comment:
This field can be retrieved from exchange future. No need to change method contract.